src/java.net.http/share/classes/java/net/http/HttpClient.java

```diff
@@ -730,4 +741,121 @@ public enum Redirect {
     public WebSocket.Builder newWebSocketBuilder() {
         throw new UnsupportedOperationException();
     }
+
+    /**
+     * Initiates an orderly shutdown in which previously submitted
+     * operations are completed, but no new request will be accepted.
+     * Invocation has no additional effect if already shut down.
+     *
+     * <p>This method does not wait for previously submitted request
+     * to complete execution.  Use {@link #awaitTermination awaitTermination}
+     * to do that.
+     *
+     * @implSpec
+     * The default implementation of this method does nothing. Subclasses should
+     * override this method to implement the appropriate behavior.
+     *
+     * @since 21
+     */
+    public void shutdown() { }
+
+    /**
+     * Blocks until all operations have completed execution after a shutdown
+     * request, or the timeout occurs, or the current thread is
+     * interrupted, whichever happens first.
+     *
+     * @implSpec
+     * The default implementation of this method does nothing and returns true.
+     * Subclasses should override this method to implement the proper behavior.
+     *
+     * @param duration the maximum time to wait
+     * @return {@code true} if this client terminated and
+     *         {@code false} if the timeout elapsed before termination
+     * @throws InterruptedException if interrupted while waiting
+     *
+     * @since 21
+     */
+    public boolean awaitTermination(Duration duration) throws InterruptedException {
+        Objects.requireNonNull(duration);
+        return true;
+    }
+
+    /**
+     * Returns {@code true} if all operations have completed following shut down.
+     * Note that {@code isTerminated} is never {@code true} unless
+     * either {@code shutdown} or {@code shutdownNow} was called first.
+     *
+     * @implSpec
+     * The default implementation of this method does nothing and returns false.
+     * Subclasses should override this method to implement the proper behavior.
+     *
+     * @return {@code true} if all tasks have completed following shut down
+     *
+     * @since 21
+     */
+    public boolean isTerminated() {
+        return false;
+    }
+
+    /**
+     * This method is called if the thread waiting on {@link #close()} is
+     * interrupted. An implementation of this method may attempt to
+     * interrupt operations that are actively running.
+     * The behavior of actively running operations when interrupted
+     * is undefined. In particular, there is no guarantee that
+     * interrupted operations will terminate, or that code waiting
+     * on these operations will ever be notified.
+     *
+     * @implSpec
+     * The default implementation of this method simply calls {@link #shutdown()}.
+     * Subclasses should override this method to implement the appropriate
+     * behavior.
+     *
+     * @since 21
+     */
+    public void shutdownNow() {
```

Comment:
The first sentence of the `shutdownNow` api doc should describe the behavior of the method itself, ie that it does an immediate shutdown.

Also, I wonder if there is another way to formulate these implSpec comments? Naive readers might think they have to override the method.

Reply:
I have added some text to the `@apiNote` and added a link in all the `@implSpec`. Does that read better?

Reply:
I think this part in this method's javadoc is a bit out of place:

> It is called if the thread waiting on {@link #close()} is interrupted.

It is just one of the place from where this method will be called from. Since the `shutdownNow()` is a public API, it could be called by applications too. Perhaps we should just remove that sentence from that method's javadoc? The javadoc of `close()` explains in a better way that this `shutdownNow()` method might get invoked from within `close()`, which I think should be enough.

Reply:
Right. In my first prototype I had made this method protected, until I realized it was actually better to make it public. This is a remnant of that first prototype. I will remove this sentence here, since as you noticed it's already specified at in close().